### PR TITLE
Remove glue table stack

### PR DESCRIPTION
**Pull request by rleighton**
## Before
Glue tables were declared via CF and generated from the registry of all log parsers. 
As the number of log parser grew we exceeded the max CF template size.
See: https://github.com/panther-labs/panther-enterprise/issues/187

Also, we are moving to a SAR deployment model which means that generating stacks is not an option. 

The solution is to create tables only when they are needed during log source onboarding.

## After

- The source-api creates tables needed for the specified log sources when they are registered and the views updated.
- The glue stack and glue stack generation has been removed.
- During deployment all _deployed_ tables are updated to get any schema changes and their partitions sync'd.
- The legacy glue stack is NOT removed during deployment to avoid disruption to log processing, it is removed during a teardown.

## Testing

- mage test:ci
- mage deploy + mage teardown
- mage deploy v1.3.0 + mage deploy master + mage teardown

### CODE REVIEW

File: tools/mage/deploy_onboard.go [hunk 1]
```diff
@@ -114,6 +114,18 @@ func registerPantherAccount(awsSession *session.Session, settings *config.Panthe
 		return fmt.Errorf("error calling lambda to register account: %v", err)
 	}
 
+	// collect the log types configured in panther_config.yml
```
Comment (rleighton) [inline tools/mage/deploy_onboard.go:117]: Moving this here and using it to compare fixes a bug where if you added Cloudtrail AFTER having onboarded then the source integration was not being updated.

File: tools/mage/deploy.go [hunk 2]
```diff
@@ -435,9 +435,15 @@ func deployMainStacks(awsSession *session.Session, settings *config.PantherConfi
 		c <- goroutineResult{summary: frontendStack, err: err}
 	}(results)
 
+	// Glue (not a stack!, moving to custom resource)
```
Comment (austinbyers) [inline tools/mage/deploy.go:438]: Is this a TODO - we need to add a custom resource?

If so, I'm a bit confused - I thought `source-api` was handling all of the glue tables? Can you help clarify:

* What does `source-api` now do with Glue tables?
* What needs to happen on every deploy with a custom resource? Is that coming in the next PR?
* What does `glue:sync` do and when should it be run?

**Comment by rleighton** [comment]
> > The legacy glue stack is NOT removed during deployment to avoid disruption to log processing, it is removed during a teardown.
> 
> So this means all existing deployments need to manually delete the glue stack? Or they should leave it indefinitely and it won't ever cause problems? Neither option seems very appealing.
> 
> My understanding was that log processing works fine without the glue tables, but historical searches wouldn't work until the glue catalog is configured correctly? If that's the case, could we check for the glue stack and, if it exists, delete it / update glue before continuing the deploy?

The existing log processing needs to interact with the glue catalog to manage partitions. Dropping a table w/out stopping log processing will cause errors.  Leaving  the stack causes no harm and will be removed on teardown.  This is the simplest and most robust "migration".

File: tools/mage/deploy.go (hunk 2 above)
Comment (rleighton) [reply tools/mage/deploy.go:438]: The source-api will add new tables now rather than adding all tables at once via CF.

The code executed above needs to be executed on on each deploy to manage any schema changes to existing tables. Before the CF took care of that. This code also to sets up Athena (the Athena part is not new). This the part I was hoping you would migrate to a custom resource. 

`glue:sync` should not be needed anymore except in cases of migration, so I want to leave it.

Comment (austinbyers) [reply tools/mage/deploy.go:438]: > The code executed above needs to be executed on on each deploy to manage any schema changes to existing tables. Before the CF took care of that. This code also sets up Athena

I see, so if an existing log schema adds a new field for a table that's already created, this code is what will update the table definition?

One caveat: custom resources will never be updated unless one of their parameters changes. You want it to trigger on every deploy, right? We could add a dummy parameter which is a random string or maybe a hash of the schema definitions

As for ownership, I'm happy to try it, but packaging is a hard requirement for the next release, so depending on timeline I may need your help.

Comment (rleighton) [reply tools/mage/deploy.go:438]: Exactly, this is needed on each deploy to deal with changes (which were handled in the CF).

I just figured since time is short and you create the other custom resources you would be able to do it most quickly (just wrap that one function). I can give it a try in this PR if you like, just may take me a little longer.

Comment (rleighton) [reply tools/mage/deploy.go:438]: I did 2min of googling, some folks just add a timestamp as a parameter to ensure the resource is invoked each time.

Comment (rleighton) [reply tools/mage/deploy.go:438]: How would that work in SAR (in mage we can just add a timestamp to the calling stack)?

Comment (rleighton) [reply tools/mage/deploy.go:438]: Also, there is this (which I hope we never have to use): https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html

Comment (rleighton) [reply tools/mage/deploy.go:438]: Actually, for SAR we could just use the Panther version, which should change each time. For mage we would set the parameter to the version+timestamp (so devs can test changes).

Comment (rleighton) [reply tools/mage/deploy.go:438]: I'll go ahead now and take a crack at the custom resource in this PR.

Comment (austinbyers) [reply tools/mage/deploy.go:438]: > I did 2min of googling, some folks just add a timestamp as a parameter to ensure the resource is invoked each time.

Makes sense. Yes, in SAR it can just be the Panther version and then mage can be whatever we want

But is this a fast operation if there are no changes? We don't want to invoke a custom resource on every deploy (for developers) if it takes several minutes but doesn't do anything. Even if it is fast, ideally we can skip the custom resource if it didn't change. That's why I had suggested a hash of the log schemas, so it only changes when the logs themselves have actually changed. That hash can be hardcoded when publishing to SAR, but I'm not sure if that's an easy thing to compute right now?

Also, will the 15-minute Lambda timeout be sufficient? If not, we'll have to use SNS custom resource invocation, which I'm not as familiar with

File: tools/mage/deploy.go [hunk 3]
```diff
@@ -450,25 +456,62 @@ func deployMainStacks(awsSession *session.Session, settings *config.PantherConfi
...
+	deployedLogTables, err := gluetables.DeployedTables(glueClient)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	for _, logTable := range deployedLogTables {
+		logger.Infof("deploy: creating/updating glue table %s.%s", logTable.DatabaseName(), logTable.TableName())
+		err := logTable.CreateOrUpdateTable(glueClient, outputs["ProcessedDataBucket"])
+		if err != nil {
+			return errors.Wrapf(err, "could not create glue log table for %s.%s", logTable.DatabaseName(), logTable.TableName())
+		}
```
Comment (kostaspap) [inline tools/mage/deploy.go:496]: I am not sure I understand, why is this needed? The tables are already deployed, why do we need to run "CreateOrUpdate" on them?

File: tools/mage/deploy.go [hunk 4]
```diff
@@ -450,25 +456,62 @@ func deployMainStacks(awsSession *session.Session, settings *config.PantherConfi
...
+		err = logTable.SyncPartitions(glueClient, s3Client, zeroStartTime)
+		if err != nil {
+			logger.Fatalf("deploy: failed syncing %s.%s: %v", logTable.DatabaseName(), logTable.TableName(), err)
+		}
```
Comment (kostaspap) [inline tools/mage/deploy.go:500]: This can be time-consuming, it could delay deployments especially during development. Seems it would be required only on specific version migrations?

File: tools/mage/deploy.go [hunk 5]
```diff
@@ -450,25 +456,62 @@ func deployMainStacks(awsSession *session.Session, settings *config.PantherConfi
...
+		logger.Infof("deploy: creating/updating glue table %s.%s",
+			ruleTable.DatabaseName(), ruleTable.TableName())
+		err = ruleTable.CreateOrUpdateTable(glueClient, outputs["ProcessedDataBucket"])
+		if err != nil {
+			return errors.Wrapf(err, "could not create glue rule table for %s.%s", ruleTable.DatabaseName(), ruleTable.TableName())
+		}
+		err = ruleTable.SyncPartitions(glueClient, s3Client, zeroStartTime)
+		if err != nil {
+			logger.Fatalf("deploy: failed syncing %s.%s: %v", ruleTable.DatabaseName(), ruleTable.TableName(), err)
+		}
```
Comment (kostaspap) [inline tools/mage/deploy.go:514]: Same as above. Seems it would be required only when migrating between specific versions.

File: tools/mage/glue_namespace.go [hunk 6]
```diff
@@ -85,16 +63,25 @@ func (t Glue) Sync() {
 		logger.Infof("syncing %s", name)
 		err := table.SyncPartitions(glueClient, s3Client, startDate)
 		if err != nil {
-			logger.Fatalf("failed syncing %s: %v", name, err)
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == glue.ErrCodeEntityNotFoundException {
```
Comment (kostaspap) [inline tools/mage/glue_namespace.go:66]: * This code seems to be pretty identical to what is running in the `deploy.deployGlue` command. Might be easy to extract to common function? 
* Line 58 should be `gluetables.DeployedTables()`, right?

File: internal/log_analysis/awsglue/integration_test.go [hunk 7]
```diff
@@ -103,15 +82,25 @@ func TestIntegrationGlueMetadataPartitions(t *testing.T) {
 		removeTables(t)
 	}()
 
-	gm := NewGlueTableMetadata(models.LogData, testTable, "test table", GlueTableHourly, &testEvent{})
+	// get the meta data, note we update the schema from the one used in setupTables()
+	gm := NewGlueTableMetadata(models.RuleData, testTable, "test table", GlueTableHourly, &testEventModified{})
```
Comment (kostaspap) [inline internal/log_analysis/awsglue/integration_test.go:86]: For my own understanding, is there a reason it was changed from `LogData` to `RuleData`?

**Comment by kostaspap** [comment]
Nice cleanup, very happy to see the CFN generation gone :D

File: internal/core/source_api/api/put_integration.go [hunk 8]
```diff
@@ -89,25 +89,25 @@ func (api API) PutIntegration(input *models.PutIntegrationInput) (*models.Source
...
-	}
-
 	switch aws.StringValue(input.IntegrationType) {
 	case models.IntegrationTypeAWS3:
 		permissionAdded, err = AllowExternalSnsTopicSubscription(*input.AWSAccountID)
 		if err != nil {
 			err = errors.Wrap(err, "Failed to add permissions to log processor queue")
 			return nil, putIntegrationInternalError
 		}
+		err = addGlueTables(input)
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/core/source_api/api/put_integration.go:99]: Good point.

File: internal/core/source_api/api/put_integration.go [hunk 9]
```diff
@@ -148,8 +148,9 @@ func (api API) integrationAlreadyExists(input *models.PutIntegrationInput) error
 				return nil
 			case models.IntegrationTypeAWS3:
 				if *existingIntegration.AWSAccountID == *input.AWSAccountID &&
-					*existingIntegration.IntegrationLabel == *input.IntegrationLabel {
-					// Log sources for same account need to have different labels
+					*existingIntegration.IntegrationLabel == *input.IntegrationLabel &&
+					len(existingIntegration.LogTypes) == len(input.LogTypes) {
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/core/source_api/api/put_integration.go:152]: Yeah, that was me "fixing" a bug , but it is not a bug. The onboarding code has a bug where if you add CT in the panther config after deploying it is not picked up.  The real fix is for that code to call update, in that case since put does not update. Will fix.

File: internal/log_analysis/awsglue/awsglue.go [hunk 10]
```diff
@@ -49,6 +49,7 @@ var (
 		LogProcessingDatabaseName: LogProcessingDatabaseDescription,
 		RuleMatchDatabaseName:     RuleMatchDatabaseDescription,
 		ViewsDatabaseName:         ViewsDatabaseDescription,
+		TempDatabaseName:          TempDatabaseDescription,
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/log_analysis/awsglue/awsglue.go:52]: Yes, used for Parquet CTAS and declared here to minimize merge changes. Originally it was not put in this list to keep it out of the UI, but we need a master list so  put it back in.

File: internal/log_analysis/awsglue/integration_test.go (hunk 7 above)
Comment (rleighton) [reply internal/log_analysis/awsglue/integration_test.go:86]: Yes, to exercise all the code paths in the test. For rule data we append some columns and I wanted to cover that.

File: internal/log_analysis/awsglue/schema.go [hunk 11]
```diff
@@ -26,13 +26,107 @@ import (
...
+			Comment: "The tags of the rule that generated this alert",
+		},
+	}
+)
+
+type Column struct {
+	Name     string
+	Type     string // this is the Glue type
+	Comment  string `json:",omitempty"`
+	Required bool   `json:"-"` // do NOT serialize! Not used for Glue types (used for doc).
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/log_analysis/awsglue/schema.go:127]: Good point, yes.

File: internal/log_analysis/gluetables/gluetables.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/glue"
+	"github.com/aws/aws-sdk-go/service/glue/glueiface"
+	"github.com/pkg/errors"
+
+	"github.com/panther-labs/panther/internal/log_analysis/awsglue"
+	"github.com/panther-labs/panther/internal/log_analysis/log_processor/registry"
+)
+
+func DeployedTables(glueClient glueiface.GlueAPI) (deployedLogTables []*awsglue.GlueTableMetadata, err error) {
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/log_analysis/gluetables/gluetables.go:31]: I wish, import cycle issues.

File: tools/mage/deploy.go [hunk 13]
```diff
@@ -435,9 +435,15 @@ func deployMainStacks(awsSession *session.Session, settings *config.PantherConfi
...
+	// Glue (not a stack!, moving to custom resource)
+	count++
+	go func(c chan goroutineResult) {
+		c <- goroutineResult{summary: "panther-glue-tables", err: deployGlue(awsSession, outputs)}
+	}(results)
+
 	// Wait for stacks to finish.
 	// There are two stacks before and one stack after
-	logResults(results, "deploy", 3, count+2, len(allStacks))
+	logResults(results, "deploy", 3, count+2, len(allStacks)+1)
```
[the remark this answers is not in the document]
Comment (rleighton) [reply tools/mage/deploy.go:446]: Moving to custom resource, so this is going to be deleted.

File: tools/mage/deploy.go (hunk 3 above)
Comment (rleighton) [reply tools/mage/deploy.go:496]: Schema change (e.g., add a Panther column).

File: tools/mage/deploy.go (hunk 4 above)
Comment (rleighton) [reply tools/mage/deploy.go:500]: I'll be computing a hash of the deployed tables, so this will only run when needed in the custom resource.

File: tools/mage/deploy.go (hunk 5 above)
Comment (rleighton) [reply tools/mage/deploy.go:514]: See above, next update should be more clever.

File: tools/mage/deploy_onboard.go [hunk 14]
```diff
@@ -126,7 +138,8 @@ func registerPantherAccount(awsSession *session.Session, settings *config.Panthe
 		}
 		if aws.StringValue(integration.AWSAccountID) == accountID &&
 			*integration.IntegrationType == models.IntegrationTypeAWS3 &&
-			*integration.IntegrationLabel == genLogProcessingLabel(awsSession) {
+			*integration.IntegrationLabel == genLogProcessingLabel(awsSession) &&
+			len(integration.LogTypes) == len(logTypes) {
```
[the remark this answers is not in the document]
Comment (rleighton) [reply tools/mage/deploy_onboard.go:142]: Yeah, will fix. The _bug_ is we need to do an update if it already exists and the log types have changed. Will fix.

File: tools/mage/glue_namespace.go (hunk 6 above)
Comment (rleighton) [reply tools/mage/glue_namespace.go:66]: I wrestled with that for a bit and moved on ... might make more sense now that the code is in better shape. I'll give it a try.

File: tools/mage/deploy_onboard.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (rleighton) [reply tools/mage/deploy_onboard.go:143]: Fixed

File: tools/mage/deploy_onboard.go [hunk 15]
```diff
@@ -185,7 +196,25 @@ func registerPantherAccount(awsSession *session.Session, settings *config.Panthe
 
 			return fmt.Errorf("error calling lambda to register account for log processing: %v", err)
 		}
-		logger.Infof("deploy: account %s registered for log processing", accountID)
+		logger.Infof("deploy: account %s registered for log processing on bucket %s for %v",
+			accountID, auditLogsBucket, logTypes)
+	}
+
+	if logProcessingIntegration != nil { // need to update
```
Comment (rleighton) [inline tools/mage/deploy_onboard.go:203]: This fixes the bug where we were not updating log types when changes were made in `panther_config.yml`

File: tools/mage/deploy.go (hunk 2 above)
Comment (rleighton) [reply tools/mage/deploy.go]: Added custom resources

File: internal/log_analysis/awsglue/table_metadata.go [hunk 16]
```diff
@@ -114,6 +118,80 @@ func (gm *GlueTableMetadata) PartitionKeys() (partitions []PartitionKey) {
...
+	_, err := glueClient.CreateTable(createTableInput)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == glue.ErrCodeAlreadyExistsException {
+			// need to do an update
+			updateTableInput := &glue.UpdateTableInput{
+				DatabaseName: &gm.databaseName,
+				TableInput:   tableInput,
+			}
+			_, err := glueClient.UpdateTable(updateTableInput)
+			return err
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/log_analysis/awsglue/table_metadata.go]: Done

File: internal/log_analysis/awsglue/table_metadata.go [hunk 17]
```diff
@@ -114,6 +118,80 @@ func (gm *GlueTableMetadata) PartitionKeys() (partitions []PartitionKey) {
...
+		}
+		return err
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/log_analysis/awsglue/table_metadata.go]: Done

File: internal/log_analysis/awsglue/wrappers.go [hunk 18]
```diff
@@ -30,16 +30,29 @@ import (
 
 // Wrapper functions to reduce boiler-plate code in callers
 
+func CreateDatabase(client glueiface.GlueAPI, databaseName, databaseDescripiton string) (*glue.CreateDatabaseOutput, error) {
```
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/log_analysis/awsglue/wrappers.go]: Done

File: internal/core/source_api/api/put_integration.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/core/source_api/api/put_integration.go]: Fixed

File: internal/log_analysis/awsglue/schema.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (rleighton) [reply internal/log_analysis/awsglue/schema.go]: Changed

File: tools/mage/deploy.go (hunk 5 above)
Comment (rleighton) [reply tools/mage/deploy.go]: Added use of table "signatures" for change detection.

File: tools/mage/glue_namespace.go (hunk 6 above)
Comment (rleighton) [reply tools/mage/glue_namespace.go:65]: Not in deploy anymore, in custom resource.

File: internal/log_analysis/awsglue/table_metadata.go [hunk 19]
```diff
@@ -114,6 +125,104 @@ func (gm *GlueTableMetadata) PartitionKeys() (partitions []PartitionKey) {
...
+				Parameters:           descriptorParameters,
+			},
+		},
+		TableType: aws.String("EXTERNAL_TABLE"),
+	}
+}
+
+func (gm *GlueTableMetadata) Signature() (string, error) {
+	tableInput := gm.glueTableInput("")
+	tableInputJSON, err := json.MarshalIndent(tableInput, "", "") // Indent forces sorting for consistency
```
Comment (rleighton) [inline internal/log_analysis/awsglue/table_metadata.go:192]: I did not use `jsoniter` here because it did not produce consistent data.

File: deployments/log_analysis.yml [hunk 20]
```diff
@@ -87,6 +93,22 @@ Conditions:
...
+    Properties:
+      AthenaResultsBucket: !Ref AthenaResultsBucket
+      ServiceToken: !Sub arn:${AWS::Partition}:lambda:${AWS::Region}:${AWS::AccountId}:function:panther-cfn-custom-resources
+
+  ###### Update Glue Table Schemas for Deployed Tables #####
+  UpdateGlueTables:
+    DependsOn: AthenaConfigure
+    Type: Custom::UpdateGlueTables
+    Properties:
+      TablesSignature: !Ref TablesSignature # this should change when tables need to be updated, for CF this can be the Panther version
```
Comment (austinbyers) [inline deployments/log_analysis.yml:108]: Nice! We can also just compute the hash as part of the release process and hardcode it in the final master template, right? Either way, the version will be available

File: deployments/master.yml [hunk 21]
```diff
@@ -110,7 +110,8 @@ Resources:
         # If the AlarmTopicArn parameter is blank, Bootstrap will create the topic.
         # That's why this uses the Bootstrap output instead of the parameter.
         AlarmTopicArn: !GetAtt Bootstrap.Outputs.AlarmTopicArn
-
+        AthenaResultsBucket: !GetAtt Bootstrap.Outputs.AthenaResultsBucket
+        ProcessedDataBucket: !GetAtt Bootstrap.Outputs.ProcessedDataBucket
```
Comment (austinbyers) [inline deployments/master.yml:114]: Yay! Thanks for remembering to update the master stack, I'll use CI to help enforce this in the future

File: internal/core/custom_resources/resources/athena.go [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+				workgroup, props.AthenaResultsBucket, err)
+		}
+
+		return resourceID, nil, nil
+	case cfn.RequestDelete: // noop
+		return event.PhysicalResourceID, nil, nil
+
+	default:
+		return "", nil, fmt.Errorf("unknown request type %s", event.RequestType)
+	}
```
Comment (austinbyers) [inline internal/core/custom_resources/resources/athena.go:56]: Since it isn't really possible to have an invalid request type, I usually just collapse these:

```suggestion
	default:
    		// ignore deletes
   		return event.PhysicalResourceID, nil, nil
```

Whichever you prefer

File: internal/core/custom_resources/resources/glue.go [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/panther-labs/panther/internal/log_analysis/athenaviews"
+	"github.com/panther-labs/panther/internal/log_analysis/awsglue"
+	"github.com/panther-labs/panther/internal/log_analysis/gluetables"
+)
+
+type UpdateGlueTablesProperties struct {
+	// TablesSignature should change every time the tables change (for SAR this can be the Panther version)
```
Comment (austinbyers) [inline internal/core/custom_resources/resources/glue.go:38]: Nit: Replace "SAR" references to "pre-packaged" or "master stack" or similar

File: internal/core/source_api/api/glue_utils.go [hunk 24]
```diff
@@ -18,3 +17,25 @@ package gluecf
...
+		}
+	}
+
+	// update the views with the new tables
+	err := athenaviews.CreateOrReplaceViews(glueClient, athenaClient)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (austinbyers) [inline internal/core/source_api/api/glue_utils.go:40]:
```suggestion
	return athenaviews.CreateOrReplaceViews(glueClient, athenaClient)
```

File: internal/log_analysis/awsglue/table_metadata.go (hunk 19 above)
Comment (austinbyers) [reply internal/log_analysis/awsglue/table_metadata.go:192]: Ah, there's a setting for that:

```go
json := jsoniter.ConfigCompatibleWithStandardLibrary
b, err := json.Marshal(m)
```

This will serialize in lexicographic order like the standard library. Not as fast as the default jsoniter speed, but still faster than the stdlib

File: tools/mage/deploy_onboard.go (hunk 15 above)
Comment (austinbyers) [reply tools/mage/deploy_onboard.go:203]: nit: can we elaborate on the comment? It looks like a leftover TODO instead of "this setting is defined, we need to update the integration settings..."

File: tools/mage/teardown.go [hunk 25]
```diff
@@ -212,6 +212,11 @@ func destroyCfnStacks(awsSession *session.Session, identity *sts.GetCallerIdenti
 		logger.Fatal(err)
 	}
 
+	// In v1.4.0 we removed the stack `panther-glue`, delete it (we can remove this after a few more releases)
+	if err := deleteStack(client, aws.String("panther-glue")); err != nil {
```
Comment (austinbyers) [inline tools/mage/teardown.go:216]: I thought you wanted to leave it for stability?

File: internal/core/custom_resources/resources/glue.go [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+
+		// update the views with the new tables
+		err = athenaviews.CreateOrReplaceViews(glueClient, athenaClient)
+		if err != nil {
+			return "", nil, errors.Wrap(err, "failed creating views")
+		}
+
+		return resourceID, nil, nil
+
+	case cfn.RequestDelete:
```
Comment (austinbyers) [inline internal/core/custom_resources/resources/glue.go:105]: Since there's deletion code in the custom resource, does that mean we can remove anything from `mage teardown`?

File: tools/mage/deploy.go (hunk 4 above)
Comment (rleighton) [reply tools/mage/deploy.go]: Done

File: deployments/log_analysis.yml (hunk 20 above)
Comment (rleighton) [reply deployments/log_analysis.yml:108]: Either way.

File: tools/mage/teardown.go (hunk 25 above)
Comment (rleighton) [reply tools/mage/teardown.go:216]: Not when we teardown, nothing is left :)

File: internal/core/custom_resources/resources/glue.go (hunk 26 above)
Comment (rleighton) [reply internal/core/custom_resources/resources/glue.go:105]: Already done. The stack is gone from the slice of stacks. Just the explicit delete is there to get rid of legacy stacks. We can probably remove that in a month or two.

File: tools/mage/teardown.go (hunk 25 above)
Comment (austinbyers) [reply tools/mage/teardown.go:216]: (I thought this was in `deploy.go` for some reason, my bad lol)
